core/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java

```diff
@@ -0,0 +1,308 @@
+        }
+        return parse(parser, valueSupplier.get(), null);
+    }
+
+    /**
+     * Parses a Value from the given {@link XContentParser}
+     * @param parser the parser to build a value from
+     * @param value the value to fill from the parser
+     * @return the parsed value
+     * @throws IOException if an IOException occurs.
+     */
+    public Value parse(XContentParser parser, Value value) throws IOException {
+        return parse(parser, value, null);
+    }
+
+    /**
+     * Parses a Value from the given {@link XContentParser}
+     * @param parser the parser to build a value from
+     * @param value the value to fill from the parser
+     * @param context an optional context that is passed along to all declared field parsers
+     * @return the parsed value
+     * @throws IOException if an IOException occurs.
+     */
+    public Value parse(XContentParser parser, Value value, Context context) throws IOException {
+        String currentFieldName = null;
+        XContentParser.Token token;
+        if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            token = parser.currentToken();
+        } else {
+            token = parser.nextToken();
+        }
+
+        FieldParser<Value> fieldParser = null;
+        if (token != XContentParser.Token.START_OBJECT) {
+            throw new IllegalStateException("[" + name  + "] Expected START_OBJECT but was: " + token);
+        }
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+                fieldParser = getParser(currentFieldName);
+            } else {
+                if (currentFieldName == null) {
+                    throw new IllegalStateException("[" + name  + "] no field found");
+                }
+                assert fieldParser != null;
+                fieldParser.assertSupports(name, token, currentFieldName, parser.getParseFieldMatcher());
+                parseNextToken(parser, fieldParser, currentFieldName, value, context);
+                fieldParser = null;
+            }
+        }
+        return value;
+    }
+
+    private void parseArray(XContentParser parser, FieldParser<Value> fieldParser, String currentFieldName, Value value, Context context) throws IOException {
+        assert parser.currentToken() == XContentParser.Token.START_ARRAY : "Token was: " + parser.currentToken();
+        parseValue(parser, fieldParser, currentFieldName, value, context);
+    }
+
+    private void parseValue(XContentParser parser, FieldParser<Value> fieldParser, String currentFieldName, Value value, Context context) throws IOException {
+        try {
+            fieldParser.parser.parse(parser, value, context);
+        } catch (Throwable ex) {
+            throw new ParsingException(new Index("_na_"), parser, "[" + name  + "] failed to parse field [" + currentFieldName + "]", ex);
```

Comment:
we remove this stuff from parsing exception in a different PR soon (query refactorign so this will go away)